metrics-scraper/monitoring.go

```diff
@@ -69,80 +70,146 @@ func Start(configPath string) {
+	if err == nil {
+		// no errors, let's start polling
+		return c, true
+	}
+
+	ticker := time.NewTicker(pollRate)
+	for {
+		select {
+		case <-ctx.Done():
+			log.Printf("job terminated for %v", config.Moniker)
+			return nil, false
+		case <-ticker.C:
+			c, err = rpc.DialContext(ctx, config.RPCURL)
+			if err != nil {
+				log.Printf("error: couldn't connect to %v at %v: %v", config.Moniker, config.RPCURL, err)
+				continue
+			}
+
+			return c, true
+		}
+	}
+}
+
 func run(
 	ctx context.Context,
 	wg *sync.WaitGroup,
 	pollRate time.Duration,
 	networkID string,
 	config NodeConfig,
 ) {
-	c, err := rpc.DialContext(ctx, config.RPCURL)
-	if err != nil {
-		log.Fatalf("error: couldn't connect to %v at %v: %v", config.Moniker, config.RPCURL, err)
+	defer wg.Done()
+
+	c, ok := tryConnect(ctx, pollRate, config)
+	if !ok {
+		// only case this happen is that the context been canceled
+		return
 	}
 
 	ticker := time.NewTicker(pollRate)
 	for {
 		select {
 		case <-ctx.Done():
-			wg.Done()
 			log.Printf("job terminated for %v", config.Moniker)
 			return
 		case <-ticker.C:
 			if err := pollData(ctx, c, config.Moniker, networkID); err != nil {
 				log.Printf("error polling data for %v: %v", config.Moniker, err)
+			} else {
+				log.Printf("data polled successfully for: %v", config.Moniker)
 			}
 		}
 	}
 }
 
 func pollData(ctx context.Context, client *rpc.Client, moniker, networkID string) error {
+	errs := []error{}
 	if err := sidecarsVersion(ctx, client, moniker, networkID); err != nil {
-		return err
+		errs = append(errs, err)
 	}
 
 	if err := nodeVersion(ctx, client, moniker, networkID); err != nil {
-		return err
+		errs = append(errs, err)
 	}
 
 	if err := latestBlockAndTimestamp(ctx, client, moniker, networkID); err != nil {
-		return err
+		errs = append(errs, err)
 	}
 
-	return nil
+	if len(errs) > 0 {
+		mezodUpGauge.WithLabelValues(moniker, networkID).Set(0)
+	} else {
+		mezodUpGauge.WithLabelValues(moniker, networkID).Set(1)
+	}
+
+	return errors.Join(errs...)
 }
 
-func latestBlockAndTimestamp(ctx context.Context, client *rpc.Client, moniker, networkID string) error {
+func latestBlockAndTimestamp(ctx context.Context, client *rpc.Client, moniker, networkID string) (err error) {
+	defer func() {
+		if err != nil {
+			// set latest block to 0 for error
+			mezoLatestBlockGauge.WithLabelValues(moniker, networkID).Set(0)
+			// set latest timestamp to 0 for error
+			mezoLatestTimestampGauge.WithLabelValues(moniker, networkID).Set(0)
+		}
+	}()
+
 	result := struct {
 		Number    string `json:"number"`
 		Timestamp string `json:"timestamp"`
 	}{}
-	err := client.CallContext(ctx, &result, ethGetBlockByNumberEndpoint, "latest", false)
+
+	err = client.CallContext(ctx, &result, ethGetBlockByNumberEndpoint, "latest", false)
 	if err != nil {
-		mezodUpGauge.WithLabelValues(moniker, networkID).Set(0)
 		return fmt.Errorf("couldn't call %v for %v: %v", ethGetBlockByNumberEndpoint, moniker, err)
```

Comment:
```suggestion
        err = fmt.Errorf("couldn't call %v for %v: %v", ethGetBlockByNumberEndpoint, moniker, err)
		return 
```

Reply:
addressed in: 94e0597d570fbb1a694bc047ed7b89d8570a1289